### PR TITLE
Ygy

### PR DESCRIPTION
**Pull request by cxjyxxme**

### CODE REVIEW

File: python/jittor/nn.py [hunk 1]
```diff
@@ -161,6 +161,14 @@ def step(self, loss):
         # sync such parameters to reduce memory consumption
         jt.sync(self.no_grad_parameters)
 
+    def sync(self):
+        ps = self.parameters
+        for p in ps:
+            temp = jt.compile_extern.nccl_ops.nccl_broadcast(p, 0)
+            p -= p
+            p += temp
```
Comment (cjld) [inline python/jittor/nn.py:169]: `p.assign(temp.detach())` seems better

File: extern/cuda/nccl/src/nccl_warper.cc [hunk 2]
```diff
@@ -8,6 +8,9 @@
 // ***************************************************************
 #include "nccl_warper.h"
 
+#ifdef HAS_CUDA
```
Comment (cjld) [inline extern/cuda/nccl/src/nccl_warper.cc:11]: remove has_cuda

File: extern/cuda/nccl/ops/nccl_reduce_op.cc [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        @if(@strcmp(@Tx,float)==0 || @strcmp(@Tx,float32)==0, ncclFloat)
+        @if(@strcmp(@Tx,int)==0 || @strcmp(@Tx,int32)==0, ncclInt)
+        @if(@strcmp(@Tx,float64)==0, ncclFloat64)
+        @if(@strcmp(@Tx,int64)==0, ncclInt64)
+    )
+    @for(i, 0, XDIM, index_t xshape@i = x->shape[@i];)
+    int size = 1 @for(i, 0, XDIM,  * xshape@{i});
+    auto* __restrict__ xp = x->ptr<Tx>();
+    auto* __restrict__ yp = y->ptr<Tx>();
+    checkCudaErrors(ncclReduce(xp, yp, size, @T_NCCL, ncclSum, root, comm, 0));
```
Comment (cjld) [inline extern/cuda/nccl/ops/nccl_reduce_op.cc:56]: 用和mpi类似的方式初始化yp
https://github.com/Jittor/jittor/pull/58/files#r410924949

File: extern/cuda/nccl/src/nccl_warper.cc [hunk 4]
```diff
@@ -26,6 +26,11 @@ nccl_initer() {
         checkCudaErrors(ncclGetUniqueId(&id));
     MPI_CHECK(MPI_Bcast((void *)&id, sizeof(id), MPI_BYTE, 0, MPI_COMM_WORLD));
     checkCudaErrors(cudaSetDevice(mpi_local_rank));
+    #ifdef HAS_CUDA
```
Comment (cjld) [inline extern/cuda/nccl/src/nccl_warper.cc:29]: 这个`HAS_CUDA`也可以去掉了， 因为编译nccl肯定有cuda
